File: frontend/src/lib/api/comparison.ts
/**
 * Typed API functions for document comparisons (Phase 12).
 *
 * Matches backend/app/schemas/comparison.py and backend/app/api/compare.py.
 *
 * Endpoints consumed:
 *   POST  /documents/compare                       → initiateComparison()
 *   GET   /documents/compare/{id}                  → getComparison()
 *   GET   /documents/compare/{id}/changes          → listComparisonChanges()
 *   GET   /documents/compare/{id}/narration        → getComparisonNarration()
 */

import { get, post } from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export type ComparisonStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED'
export type ChangeType = 'ADDED' | 'REMOVED' | 'MODIFIED'
export type ChangeSeverity = 'MAJOR' | 'MODERATE' | 'MINOR'
export type VersionState = 'CURRENT' | 'SUPERSEDED' | 'SCHEDULED'

/** Summary counts from the JSONB summary column (populated on COMPLETED). */
export interface ComparisonSummary {
  total: number
  major: number
  moderate: number
  minor: number
  alignment_degraded: boolean
}

/** Full comparison row — from GET /documents/compare/{id}. */
export interface ComparisonResponse {
  id: string
  organization_id: string
  document_a_version_id: string
  document_b_version_id: string
  /** Phase 15 — parent document IDs for "View source" deep links. */
  document_a_id: string | null
  document_b_id: string | null
  status: ComparisonStatus
  summary: ComparisonSummary | null
  error_message: string | null
  requested_by: string
  created_at: string
  completed_at: string | null
}

/** Response from POST /documents/compare — adds the `created` flag. */
export interface ComparisonCreateResponse extends ComparisonResponse {
  /** True when a new comparison was triggered; False when reusing an existing one. */
  created: boolean
}

/** Resolved provenance for one side of a change (backend SourceRef). */
export interface ComparisonSourceRef {
  document_id: string
  document_version_id: string
  document_name: string
  version_number: number
  page_number: number
  section: string | null
}

/** One detected change from GET /documents/compare/{id}/changes. */
export interface ComparisonChange {
  id: string
  comparison_id: string
  change_type: ChangeType
  severity: ChangeSeverity
  section: string | null
  old_chunk_id: string | null
  new_chunk_id: string | null
  old_text: string | null
  new_text: string | null
  truncated: boolean
  created_at: string
  old_source: ComparisonSourceRef | null
  new_source: ComparisonSourceRef | null
}

/** Paginated changes response. */
export interface ComparisonChangesResponse {
  comparison_id: string
  total: number
  items: ComparisonChange[]
}

/** LLM narration response. */
export interface ComparisonNarrationResponse {
  comparison_id: string
  narration: string
  llm_narrated: boolean
}

// ─── Request types ─────────────────────────────────────────────────────────────

export interface CompareRequest {
  document_a_version_id: string
  document_b_version_id: string
}

export interface ListChangesParams {
  severity?: ChangeSeverity
  section?: string
}

// ─── API functions ─────────────────────────────────────────────────────────────

/**
 * Initiate a comparison between two document versions.
 *
 * Idempotent — returns the existing comparison if one already exists for the
 * same pair.  Poll `getComparison()` until `status === 'COMPLETED'`.
 *
 * Returns 202 Accepted on first creation; the `created` field distinguishes
 * new vs. reused.
 */
export async function initiateComparison(
  body: CompareRequest,
): Promise<ComparisonCreateResponse> {
  return post<ComparisonCreateResponse>('/documents/compare', body)
}

/**
 * Get comparison status + summary.
 * Poll this until `status === 'COMPLETED'` or `status === 'FAILED'`.
 */
export async function getComparison(comparisonId: string): Promise<ComparisonResponse> {
  return get<ComparisonResponse>(`/documents/compare/${comparisonId}`)
}

/**
 * List detected changes for a comparison.
 * Only meaningful once status is 'COMPLETED'.
 */
export async function listComparisonChanges(
  comparisonId: string,
  params?: ListChangesParams,
): Promise<ComparisonChangesResponse> {
  const query = new URLSearchParams()
  if (params?.severity) query.set('severity', params.severity)
  if (params?.section) query.set('section', params.section)
  const qs = query.toString()
  return get<ComparisonChangesResponse>(
    `/documents/compare/${comparisonId}/changes${qs ? `?${qs}` : ''}`,
  )
}

/**
 * Get an LLM-generated narration of the comparison changes.
 * Only available when status is 'COMPLETED'.
 */
export async function getComparisonNarration(
  comparisonId: string,
): Promise<ComparisonNarrationResponse> {
  return get<ComparisonNarrationResponse>(
    `/documents/compare/${comparisonId}/narration`,
  )
}

File: backend/app/schemas/comparison.py
"""
Pydantic v2 schemas for the document comparison API (Phase 12).

Request/response models for:
  - POST  /documents/compare           — initiate or reuse a comparison
  - GET   /documents/compare/{id}      — comparison status + summary
  - GET   /documents/compare/{id}/changes — paginated list of detected changes
  - GET   /documents/compare/{id}/narration — LLM-generated narrative

All response models use from_attributes = True (ORM mode).
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class _OrmBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# ── Request ────────────────────────────────────────────────────────────────────

class CompareRequest(BaseModel):
    """Request body for POST /documents/compare."""
    document_a_version_id: str = Field(
        description="ID of the first document version (A side). The pair will be "
                    "normalized so the same two versions always map to the same "
                    "comparison row regardless of order."
    )
    document_b_version_id: str = Field(
        description="ID of the second document version (B side)."
    )


# ── Change response ────────────────────────────────────────────────────────────

class SourceRef(BaseModel):
    """Resolved provenance for one side of a change (plan §13).

    Populated by the service layer from the chunk FK join (chunk → page →
    section → version → document) — never stored on comparison_changes.
    ``None`` when the underlying chunk was deleted (FK SET NULL) — the
    denormalized old_text/new_text snapshot still renders.
    """
    document_id: str
    document_version_id: str
    document_name: str
    version_number: int
    page_number: int
    section: Optional[str] = None


class ComparisonChangeResponse(_OrmBase):
    """One detected change row (from comparison_changes)."""
    id: str
    comparison_id: str
    change_type: str            # ADDED | REMOVED | MODIFIED
    severity: str               # MAJOR | MODERATE | MINOR
    section: Optional[str] = None
    old_chunk_id: Optional[str] = None
    new_chunk_id: Optional[str] = None
    old_text: Optional[str] = None
    new_text: Optional[str] = None
    truncated: bool = False
    created_at: datetime
    old_source: Optional[SourceRef] = None
    new_source: Optional[SourceRef] = None


# ── Comparison summary (embedded in ComparisonResponse) ───────────────────────

class ComparisonSummary(BaseModel):
    """Counts extracted from the JSONB summary column."""
    total: int = 0
    major: int = 0
    moderate: int = 0
    minor: int = 0
    alignment_degraded: bool = False


# ── Comparison response ────────────────────────────────────────────────────────

class ComparisonResponse(_OrmBase):
    """Full comparison metadata (status + summary)."""
    id: str
    organization_id: str
    document_a_version_id: str
    document_b_version_id: str
    # Phase 15 — resolved parent document IDs (document_versions.document_id)
    # so the UI can build "View source" deep links without an extra lookup.
    document_a_id: Optional[str] = None
    document_b_id: Optional[str] = None
    status: str
    summary: Optional[ComparisonSummary] = None
    error_message: Optional[str] = None
    requested_by: str
    created_at: datetime
    completed_at: Optional[datetime] = None


class ComparisonCreateResponse(ComparisonResponse):
    """Response from POST /documents/compare — adds the `created` flag."""
    created: bool = Field(
        description="True when a new comparison was triggered; "
                    "False when an existing comparison was reused."
    )


# ── Changes list response ──────────────────────────────────────────────────────

class ComparisonChangesResponse(BaseModel):
    """Paginated list of comparison_changes rows."""
    comparison_id: str
    total: int
    items: list[ComparisonChangeResponse]


# ── Narration response ─────────────────────────────────────────────────────────

class ComparisonNarrationResponse(BaseModel):
    """LLM-generated narrative of the comparison changes."""
    comparison_id: str
    narration: str
    # True when narration was generated from the LLM; False for the plain-text fallback
    llm_narrated: bool = True
